flink-core/src/main/java/org/apache/flink/api/common/TaskInfo.java

```diff
@@ -0,0 +1,93 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.api.common;
+
+import com.google.common.base.Preconditions;
+
+/**
+ * Encapsulates task-specific information: name, index of subtask, parallelism and attempt number.
+ */
+public class TaskInfo {
+
+	private final String taskName;
+	private final String taskNameWithSubtasks;
+	private final int indexOfSubtask;
+	private final int numberOfParallelSubtasks;
+	private final int attemptNumber;
+
+	public TaskInfo(String taskName, int indexOfSubtask, int numberOfParallelSubtasks, int attemptNumber) {
+		Preconditions.checkArgument(indexOfSubtask >= 0, "Task index must be a non-negative number.");
+		Preconditions.checkArgument(numberOfParallelSubtasks >= 1, "Parallelism must be a positive number.");
+		Preconditions.checkArgument(indexOfSubtask < numberOfParallelSubtasks, "Task index must be less than parallelism.");
+		Preconditions.checkArgument(attemptNumber >= 0, "Attempt number must be a non-negative number.");
```

Comment:
Good set of checks. Would be good to add a non-null check for `taskName` as well.
Optional: statically importing ´checkArgument` makes it even nicer.